modules/gdscript/tests/test_completion.h

```diff
@@ -0,0 +1,200 @@
+#include "tests/test_macros.h"
+
+#include "editor/editor_settings.h"
+#include "scene/theme/theme_db.h"
+
+namespace GDScriptTests {
+
+bool match_option(Dictionary expected, ScriptLanguage::CodeCompletionOption got) {
+	if (expected.get("display", got.display) != got.display) {
+		return false;
+	}
+	if (expected.get("insert_text", got.insert_text) != got.insert_text) {
+		return false;
+	}
+	if (expected.get("kind", got.kind) != Variant(got.kind)) {
+		return false;
+	}
+	if (expected.get("location", got.location) != Variant(got.location)) {
+		return false;
+	}
+	return true;
+}
+
+void to_dict_list(Variant p_variant, List<Dictionary> &p_list) {
+	if (p_variant.is_array()) {
+		Array arr = static_cast<Array>(p_variant);
+		while (arr.size() > 0) {
+			Variant E = arr.pop_front();
+			if (E.get_type() == Variant::DICTIONARY) {
+				p_list.push_back(static_cast<Dictionary>(E));
+			}
+		}
+	}
+}
+
+void test_directory(const String &p_dir) {
+	Error err = OK;
+	Ref<DirAccess> dir(DirAccess::open(p_dir, &err));
+
+	if (err != OK) {
+		FAIL("Invalid test directory.");
+		return;
+	}
+
+	String path = dir->get_current_dir();
+
+	dir->list_dir_begin();
+	String next = dir->get_next();
+
+	while (!next.is_empty()) {
+		if (dir->current_is_dir()) {
+			if (next == "." || next == "..") {
+				next = dir->get_next();
+				continue;
+			}
+			test_directory(path.path_join(next));
+		} else {
+			if (next.ends_with(".gd")) {
+				Ref<FileAccess> acc = FileAccess::open(path.path_join(next), FileAccess::READ, &err);
+
+				if (err != OK) {
+					next = dir->get_next();
+					continue;
+				}
+
+				String code = acc->get_as_utf8_string();
+				// Require pointer sentinel char in scripts.
+				CHECK(code.find_char(0xFFFF) != -1);
+
+				ConfigFile conf;
+				if (conf.load(path.path_join(next.get_basename() + ".cfg")) != OK) {
+					FAIL("No config file found.");
+				}
+
+#ifndef MODULE_MONO_ENABLED
+				if (conf.get_value("input", "cs", false)) {
+					next = dir->get_next();
+					continue;
+				}
+#endif
+
+				EditorSettings::get_singleton()->set_setting("text_editor/completion/use_single_quotes", conf.get_value("input", "use_single_quotes", false));
+
+				List<Dictionary> include;
+				to_dict_list(conf.get_value("result", "include", Array()), include);
+
+				List<Dictionary> exclude;
+				to_dict_list(conf.get_value("result", "exclude", Array()), exclude);
+
+				List<ScriptLanguage::CodeCompletionOption> options;
+				String call_hint;
+				bool forced;
+
+				Node *owner = nullptr;
+				if (dir->file_exists(next.get_basename() + ".tscn")) {
+					String project_path = "res://completion";
+					Ref<PackedScene> scene = ResourceLoader::load(project_path.path_join(next.get_basename() + ".tscn"), "PackedScene");
+					if (scene.is_valid()) {
+						owner = scene->instantiate();
+					}
+				}
+
+				GDScriptLanguage::get_singleton()->complete_code(code, path.path_join(next), owner, &options, forced, call_hint);
+				String contains_excluded = "";
```

Comment:
```suggestion
				String contains_excluded;
```
Unnecessary default declaration